torch/testing/_internal/common_methods_invocations.py

```diff
@@ -9246,7 +9250,23 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                     assert_autodiffed=True,
                     supports_forward_ad=True,
                     supports_fwgrad_bwgrad=True,
-                    supports_two_python_scalars=True),
+                    supports_two_python_scalars=True,
+                    supports_sparse=True,
+                    supports_sparse_csr=True,
+                    supports_sparse_csc=True,
+                    supports_sparse_bsr=True,
+                    supports_sparse_bsc=True,
+                    sample_inputs_sparse_coo_func=partial(sample_inputs_elementwise_binary_operation_sparse,
```

Comment:
Only related to this PR: it seems the logic for defining support for sparse and setting the corresponding sample inputs function is a little redundant. Meaning, you need to set `supports_sparse=True` *and* pass a `sample_inputs_sparse_coo_func`. Shouldn't the latter be enough?

Internally one can use `self.sample_inputs_sparse_coo_func is not None` over `self.supports_sparse`. For BC, one could also add `self.supports_sparse = self.sample_inputs_sparse_coo_func is not None` to the constructor.

Reply:
IIUC, while `supports_sparse=True` and defining `sample_inputs_sparse_coo_func` have overlapping meanings, then just specifying `supports_sparse=True` makes also sense: in principle some sparse tests (say, conversion tests) may use strided samples.

Otherwise, I agree, defining `sample_inputs_sparse_coo_func` could imply `supports_sparse=True`.

Reply:
Fair enough. Maybe an improvement could be to have `supports_sparse: Optional[bool] = None` as parameter and do

```python
if supports_sparse is None:
    supports_sparse = sample_inputs_sparse_coo_func is not None
```

This way one can set `supports_sparse=True` if needs be, but otherwise it will be implied by the presence of a sample inputs function.

Reply:
Done: for unspecified `supports_sparse`, specifying `sample_inputs_sparse_coo_func` now implies `supports_sparse=True`